frame/ethereum/src/lib.rs

```diff
@@ -290,7 +290,32 @@ pub mod pallet {
 				"pre log already exists; block is invalid",
 			);
 
-			Self::apply_validated_transaction(source, transaction)
+			let gas_limit: u64 = match transaction {
+				Transaction::Legacy(ref t) => t.gas_limit.unique_saturated_into(),
+				Transaction::EIP2930(ref t) => t.gas_limit.unique_saturated_into(),
+				Transaction::EIP1559(ref t) => t.gas_limit.unique_saturated_into(),
+			};
+
+			let weight_limit =
+				match <T as pallet_evm::Config>::GasWeightMapping::gas_to_weight(gas_limit, true) {
+					weight_limit if weight_limit.proof_size() > 0 => {
+						// Try to subtract the encoded extrinsic from the Weight proof_size limit or fail
+						// Validate the weight limit can afford recording transaction len
+						let _ = weight_limit
+							.proof_size()
+							.checked_sub(Self::transaction_len(&transaction))
+							.ok_or(DispatchErrorWithPostInfo {
+								post_info: PostDispatchInfo {
+									actual_weight: None,
+									pays_fee: Pays::No,
+								},
+								error: sp_runtime::DispatchError::Exhausted,
```

Comment:
Can we put the verification into the `validate_transaction_in_pool`?  In this way, you do not need to add another `weight_limit` input params to the `apply_validated_transaction`, we can handle the `checked_sub(transaction_len)` in the `execute()` function together with `transaction_len`.

```rs
let transaction_len = if weight_limit.is_some() {
			Some(Self::transaction_len(transaction))
		} else {
			None
		};
```

Reply:
> Do we want to be able to keep the possibility for overriding this weight_info? 

I think so, frontier should  be generic and future proof